### PR TITLE
02 fulfillment type impacted plugins

### PR DESCRIPTION
**Pull request by sujithvn**
# This PR is replaced by https://github.com/reactioncommerce/reaction/pull/6615

Signed-off-by: Sujith mail.sujithvn@gmail.com

Resolves #6471 
Impact: **breaking**
Type: **feature**

## Issue

In the current system, the only way for order fulfillment is via shipping which is hardcoded into the system. We need to introduce the flexibility for users to add their own fulfillment types via plugins.

## Solution

We are creating a base fulfillment plugin which would enable other fulfillment types (like shipping, pickup, digital) to be introduced via plugins. Each of the newly introduced fulfillment type plugin would need to have the specific fulfillment methods also to be added as separate plugins. Example, fulfillment type 'pickup' could have fulfillment methods like 'store pickup' and 'curb-side pickup'.

This PR is 4th entry (02 fulfillment-type-impacted-plugins)

## PR Details in order

```
feat/fulfillment-types
 | 
 | __ 01 fulfillment base #6570 
 |        | __ 01-01 fulfillment-type-shipping       #6579
 |        | __ 01-02 fulfillment-type-pickup.    #6577 
 |__02 fulfillment-type-impacted-plugins. [carts, products, orders, catalogs]        <--- This PR #6582 
 |__03 placeOrder refactor #6583 

```
## Existing PRs based on the #6480 (to be updated to point to #6570) 
We shall change the base branch of the 3 existing PRs (i18n #6545, dataMigration #6544, setDefaultFFtype #6543) to  point to 01 fulfillment base

##### CARTS
* New field - cartVersion in the cart to identify the new carts. Old carts will not have this field
* New field - selectedFulfillmentType: This field allows UI to send the fulfillment-type selected by user for the product while adding to the cart. This is optional and if not selected the item would default to undecided fulfillment group.
* New field - supportedFulfillmentTypes: This field was supposed to be existing but was missing. Added now (bug fix).
* New Mutation - setFulfillmentTypeForItems: Accepts an array of itemIds and a selected fulfillment type along with cart details. This mutation would move all the provided items to the requested fulfillment group depending on conditions like FFtype is supported by product, FFtype is active etc.
* Exported ShippingMethod & Shipment from SimpleSchema - these schemas will be extended in the api-plugin-fulfillment plugin to have the allowedValues changed from the hard coded "shipping" to dynamic array containing all the installed FFTypes
* New field - fulfillmentMethod - This is field will be non-editable and used for matching the fulfillment method from other functions/calls. Matching on existing fields like 'name' which are user-input in flat-rate method is not reliable. 
* New field - displayMessageMethod - This new field will hold any string to be displayed along the FF Method in the UI, example store pickup requesting user to come in only after 24hrs of ordering.
* New field - methodAdditionalData: This could be fulfillment specific information used for processing in the system. Example Store may provide store details.
* updateCartFulfillmentGroups: logic updated - when new item is added to cart, if "selectedFulfillmentType" is provided, we add to that group, else the item goes to "Undecided" group.

##### CATALOGS
* supportedFulfillmentTypes - this field will be extended in the api-plugin-fulfillment plugin to have the allowedValues changed from the hard coded "shipping" to dynamic array containing all the installed FFTypes

##### PRODUCTS
* supportedFulfillmentTypes - this field will be extended in the api-plugin-fulfillment plugin to have the allowedValues changed from the hard coded "shipping" to dynamic array containing all the installed FFTypes

##### ORDERS
* Exported CommonOrder, orderFulfillmentGroupInputSchema, SelectedFulfillmentOption - these schemas will be extended in the api-plugin-fulfillment plugin to have the allowedValues changed from the hard coded "shipping" to dynamic array containing all the installed FFTypes
* New field - methodAdditionalData: This could be fulfillment specific information used for processing in the system. Example Store may provide store details.
(above 2 changes are spread across multiple files)

##### Integration Tests updated
* Updated the integration tests to match the changes in plugins.

## Breaking changes

Since this is the new way of handling fulfillment, the existing shipment plugins would not work along with this. User would need to migrate the data (migration script will be developed) and remove the shipment plugins (api-plugin-shipments and api-plugin-shipments-flat-rate).

## Testing
All this is new code and at this stage this is not integrated to reaction, meaning the plugins.json is not updated to include this plugin. Hence proper testing can be carried out only after merging in the remaining PRs.

### CODE REVIEW

**Comment by brent-hoover** [comment]
@sujithvn Broken build on this as well? Is this expected?

File: packages/api-plugin-carts/src/util/updateCartFulfillmentGroups.js [hunk 1]
```diff
@@ -1,61 +1,124 @@
...
  * @param {Object} context App context
  * @param {Object} cart The cart, to be mutated
  * @returns {undefined}
  */
-export default function updateCartFulfillmentGroups(context, cart) {
+export default async function updateCartFulfillmentGroups(context, cart) {
   // Every time the cart is updated, create any missing fulfillment groups as necessary.
   // We need one group per type per shop, containing only the items from that shop.
   // Also make sure that every item is assigned to a fulfillment group.
+  const { collections } = context;
```
Comment (brent-hoover) [inline packages/api-plugin-carts/src/util/updateCartFulfillmentGroups.js:80]: Combine these destructuring statements

File: pnpm-lock.yaml [hunk 2]
```diff
@@ -245,7 +245,7 @@ importers:
       '@reactioncommerce/logger': link:../../packages/logger
       '@reactioncommerce/nodemailer': 5.0.5
       '@reactioncommerce/random': link:../../packages/random
-      '@snyk/protect': 1.1032.0
```
Comment (brent-hoover) [inline pnpm-lock.yaml:248]: You should remove this (Snyk), it's going to just keep causing problems

File: packages/api-plugin-carts/src/mutations/transformAndValidateCart.js [hunk 3]
```diff
@@ -15,7 +15,7 @@ const logCtx = { name: "cart", file: "transformAndValidateCart" };
  */
 export default async function transformAndValidateCart(context, cart) {
   const { simpleSchemas: { Cart: cartSchema } } = context;
-  updateCartFulfillmentGroups(context, cart);
+  await updateCartFulfillmentGroups(context, cart);
```
Comment (brent-hoover) [inline packages/api-plugin-carts/src/mutations/transformAndValidateCart.js:18]: A little concerned about this. @aldeed was there a reason we didn't await these transforms?

**Review by brent-hoover — CHANGES_REQUESTED** [review]
A couple more changes

File: packages/api-plugin-carts/src/mutations/transformAndValidateCart.js (hunk 3 above)
Comment (sujithvn) [reply packages/api-plugin-carts/src/mutations/transformAndValidateCart.js:18]: Earlier we did not have any async calls in the `updateCartFulfillmentGroups`. Now we have added a new async call to check if the fulfillment-type selected is currently enabled/disabled. That is why I added a new await here now.

File: pnpm-lock.yaml (hunk 2 above)
Comment (sujithvn) [reply pnpm-lock.yaml:248]: removed

File: packages/api-plugin-carts/src/util/updateCartFulfillmentGroups.js (hunk 1 above)
Comment (sujithvn) [reply packages/api-plugin-carts/src/util/updateCartFulfillmentGroups.js:80]: done

**Comment by sujithvn** [comment]
This PR is replaced by https://github.com/reactioncommerce/reaction/pull/6615

**Pull request closed without merging**